File: api/src/services/versionKeys.ts
/**
 * Canonical, self-contained version normalisation for the SBOM dependency index.
 *
 * ╔══════════════════════════════════════════════════════════════════════════╗
 * ║  PARITY CONTRACT                                                          ║
 * ║  This module MUST produce byte-identical sort keys and identical          ║
 * ║  classification to container/version_keys.py. Both are validated against  ║
 * ║  the shared golden vectors in testdata/version-key-vectors.json. Never    ║
 * ║  change one side without the other and without extending the vectors.     ║
 * ╚══════════════════════════════════════════════════════════════════════════╝
 *
 * A sort key is an ASCII string whose lexicographic (code-point) ordering equals
 * the semantic ordering of release versions, so range predicates become plain
 * string range scans in BigQuery. See container/version_keys.py for the full
 * design rationale. The two implementations are intentionally hand-rolled (not
 * library-backed) so they can be guaranteed byte-identical.
 */

// --- canonical constants (MUST match version_keys.py) ---
export const ARITY = 6;
export const WIDTH = 10;
export const SEG_MAX = 10 ** WIDTH - 1;
export const PRE_NUM_WIDTH = 12;

export const EXACT_SCHEMES = new Set([
	"npm",
	"pypi",
	"gem",
	"cargo",
	"golang",
	"nuget",
	"composer",
	"pub",
	"swift",
]);
export const NON_PACKAGE_ECOSYSTEMS = new Set(["github", "githubactions"]);

const RANGE_PREFIX = new Set(["^", "~", ">", "<"]);
const GO_PSEUDO_RE = /-\d{12,14}-[0-9a-f]{12}$/;
const SHA_RE = /^[0-9a-fA-F]{7,40}$/;
const LEADING_NUM_RE = /^(\d+(?:\.\d+)*)/;
const PRE_SPLIT_RE = /[.\-_]/;

export interface NormalizedVersion {
	version_kind:
		| "release"
		| "prerelease"
		| "range"
		| "sha"
		| "branch"
		| "wildcard"
		| "empty"
		| "unparseable";
	comparable: boolean;
	exact_scheme: boolean;
	is_prerelease: boolean;
	epoch: number;
	version_sort_key: string | null;
}

/** Lenient percent-decode matching Python urllib.parse.unquote for ASCII %XX. */
export function unquote(s: string): string {
	return s.replace(/%[0-9A-Fa-f]{2}/g, (m) => String.fromCharCode(parseInt(m.slice(1), 16)));
}

const isDigits = (s: string): boolean => /^\d+$/.test(s);

function result(
	kind: NormalizedVersion["version_kind"],
	comparable: boolean,
	exactScheme: boolean,
	opts: { isPrerelease?: boolean; epoch?: number; sortKey?: string | null } = {},
): NormalizedVersion {
	return {
		version_kind: kind,
		comparable,
		exact_scheme: exactScheme,
		is_prerelease: opts.isPrerelease ?? false,
		epoch: opts.epoch ?? 0,
		version_sort_key: opts.sortKey ?? null,
	};
}

function isRangeOrWildcard(v: string): boolean {
	let s = v.trim();
	if (s.startsWith("=")) s = s.slice(1).trim();
	if (!s) return false;
	if (RANGE_PREFIX.has(s[0])) return true;
	if (v.includes("||") || v.includes(",") || v.includes(" - ")) return true;
	if (s.includes("*")) return true;
	for (const seg of s.split(PRE_SPLIT_RE)) {
		if (seg === "x" || seg === "X") return true;
	}
	return false;
}

function isSha(v: string): boolean {
	if (GO_PSEUDO_RE.test(v)) return true;
	if (isDigits(v)) return false;
	return SHA_RE.test(v);
}

function splitPre(s: string): string[] {
	return s
		.split(PRE_SPLIT_RE)
		.filter((t) => t !== "")
		.map((t) => t.toLowerCase());
}

type Parsed = { epoch: number; release: number[]; isPre: boolean; preTags: string[] } | null;

function parseSemverFamily(v: string): Parsed {
	const core = v.split("+", 1)[0];
	const dashIdx = core.indexOf("-");
	const rel = dashIdx === -1 ? core : core.slice(0, dashIdx);
	const pre = dashIdx === -1 ? "" : core.slice(dashIdx + 1);
	const release: number[] = [];
	for (const seg of rel.split(".")) {
		if (!isDigits(seg)) return null;
		release.push(parseInt(seg, 10));
	}
	if (release.length === 0) return null;
	const isPre = dashIdx !== -1 && pre !== "";
	return { epoch: 0, release, isPre, preTags: isPre ? splitPre(pre) : [] };
}

function parsePypi(v: string): Parsed {
	let s = v.toLowerCase().trim();
	let epoch = 0;
	const bang = s.indexOf("!");
	if (bang !== -1) {
		const e = s.slice(0, bang);
		epoch = isDigits(e) ? parseInt(e, 10) : 0;
		s = s.slice(bang + 1);
	}
	s = s.split("+", 1)[0];
	const m = LEADING_NUM_RE.exec(s);
	if (!m) return null;
	const release = m[1].split(".").map((x) => parseInt(x, 10));
	const rest = s.slice(m[0].length);
	let isPre = false;
	let preTags: string[] = [];
	if (rest && /^[._-]?(a|b|c|rc|alpha|beta|pre|preview|dev)/.test(rest)) {
		isPre = true;
		preTags = splitPre(rest);
	}
	return { epoch, release, isPre, preTags };
}

function parseGem(v: string): Parsed {
	const release: number[] = [];
	const preSegs: string[] = [];
	let isPre = false;
	for (const seg of v.split(".")) {
		if (isDigits(seg) && !isPre) {
			release.push(parseInt(seg, 10));
		} else {
			isPre = true;
			preSegs.push(seg);
		}
	}
	if (release.length === 0) return null;
	return { epoch: 0, release, isPre, preTags: preSegs.length ? splitPre(preSegs.join(".")) : [] };
}

function parseMaven(v: string): Parsed {
	const m = LEADING_NUM_RE.exec(v);
	if (!m) return null;
	const release = m[1].split(".").map((x) => parseInt(x, 10));
	const rest = v.slice(m[0].length).toLowerCase();
	const isPre = /(alpha|beta|rc|m\d|snapshot|cr\d|milestone|preview)/.test(rest);
	return { epoch: 0, release, isPre, preTags: isPre ? splitPre(rest) : [] };
}

function parseVersionByEco(eco: string, v: string): Parsed {
	if (["npm", "cargo", "golang", "nuget", "composer", "pub", "swift"].includes(eco)) {
		return parseSemverFamily(v);
	}
	if (eco === "pypi") return parsePypi(v);
	if (eco === "gem") return parseGem(v);
	if (eco === "maven") return parseMaven(v);
	const parsed = parseSemverFamily(v);
	if (parsed !== null) return parsed;
	return parsePypi(v);
}

export function buildSortKey(
	epoch: number,
	release: number[],
	isPrerelease: boolean,
	preTags: string[],
): string {
	const epochPart = String(Math.min(Math.max(epoch, 0), 9999)).padStart(4, "0");
	const segs: string[] = [];
	for (let i = 0; i < ARITY; i++) {
		let val = i < release.length ? release[i] : 0;
		val = Math.min(Math.max(val, 0), SEG_MAX);
		segs.push(String(val).padStart(WIDTH, "0"));
	}
	const releasePart = segs.join(".");
	let disc: string;
	if (!isPrerelease) {
		disc = "~";
	} else {
		const norm: string[] = [];
		for (const t of preTags) {
			if (isDigits(t)) norm.push("0" + t.padStart(PRE_NUM_WIDTH, "0"));
			else norm.push("1" + t);
		}
		disc = norm.length ? "-" + norm.join(".") : "-";
	}
	return `${epochPart}!${releasePart}.${disc}`;
}

export function normalizeVersion(
	ecosystem: string,
	rawVersion: string | null | undefined,
): NormalizedVersion {
	const eco = (ecosystem || "").toLowerCase();
	const exactScheme = EXACT_SCHEMES.has(eco);
	const v = unquote(rawVersion ?? "").trim();

	if (v === "") return result("empty", false, exactScheme);
	if (eco === "github") return result("branch", false, false);
	if (eco === "githubactions") {
		return result(isRangeOrWildcard(v) ? "wildcard" : "branch", false, false);
	}
	if (isRangeOrWildcard(v)) return result("range", false, exactScheme);

	let vv = v;
	if (vv.length > 1 && (vv[0] === "v" || vv[0] === "V") && /\d/.test(vv[1])) vv = vv.slice(1);
	if (vv.startsWith("=")) vv = vv.slice(1).trim();

	if (isSha(vv)) return result("sha", false, exactScheme);

	const parsed = parseVersionByEco(eco, vv);
	if (parsed === null) return result("unparseable", false, exactScheme);

	const key = buildSortKey(parsed.epoch, parsed.release, parsed.isPre, parsed.preTags);
	return result(parsed.isPre ? "prerelease" : "release", true, exactScheme, {
		isPrerelease: parsed.isPre,
		epoch: parsed.epoch,
		sortKey: key,
	});
}

export function parsePurl(
	purl: string,
): { ecosystem: string; packageName: string; version: string } | null {
	if (!purl || !purl.startsWith("pkg:")) return null;
	let body = purl.slice(4);
	body = body.split("#", 1)[0].split("?", 1)[0];
	const atIdx = body.lastIndexOf("@");
	let namePart: string;
	let version: string;
	if (atIdx !== -1) {
		namePart = body.slice(0, atIdx);
		version = body.slice(atIdx + 1);
	} else {
		namePart = body;
		version = "";
	}
	const slashIdx = namePart.indexOf("/");
	const eco = unquote(slashIdx === -1 ? namePart : namePart.slice(0, slashIdx)).toLowerCase();
	const packageName = unquote(slashIdx === -1 ? "" : namePart.slice(slashIdx + 1));
	return { ecosystem: eco, packageName, version: unquote(version) };
}

export function packageKey(ecosystem: string, packageName: string): string {
	const eco = (ecosystem || "").toLowerCase();
	const name = packageName || "";
	if (eco === "pypi") return name.replace(/[-_.]+/g, "-").toLowerCase();
	return name.toLowerCase();
}

// ───────────────────────── range predicate parser (query side) ─────────────────────────

export type RangeComparator = { op: ">=" | ">" | "<" | "<=" | "="; key: string };

/** Parse up to 3 numeric segments from a token, treating x/X/* as absent. */
function partsOf(token: string): number[] {
	const out: number[] = [];
	for (const seg of token.split(".")) {
		if (seg === "x" || seg === "X" || seg === "*" || seg === "") break;
		if (!isDigits(seg)) break;
		out.push(parseInt(seg, 10));
	}
	return out;
}

const keyOf = (parts: number[]): string => buildSortKey(0, parts, false, []);

/** Expand a caret/tilde/bare token into [lo, hiExclusive] release tuples. */
function expandRange(prefix: string, token: string): [number[], number[]] | null {
	const p = partsOf(token);
	if (p.length === 0) return null;
	const [maj = 0, min = 0, pat = 0] = p;
	const lo = [maj, min, pat];
	let hi: number[];
	if (prefix === "^") {
		if (maj > 0 || p.length === 1) hi = [maj + 1, 0, 0];
		else if (min > 0 || p.length === 2) hi = [maj, min + 1, 0];
		else hi = [maj, min, pat + 1];
	} else if (prefix === "~") {
		if (p.length >= 2) hi = [maj, min + 1, 0];
		else hi = [maj + 1, 0, 0];
	} else {
		// bare partial: "1" -> >=1.0.0 <2.0.0 ; "1.2" -> >=1.2.0 <1.3.0
		if (p.length === 1) hi = [maj + 1, 0, 0];
		else if (p.length === 2) hi = [maj, min + 1, 0];
		else return null; // full 3-part bare is exact, handled by caller
	}
	return [lo, hi];
}

export class VersionRangeError extends Error {}

/**
 * Parse an npm-style version range predicate into a flat list of comparators
 * over the sort key (ANDed). Supported grammar:
 *   <2  <=2.1  >1.0  >=1.2.3  =1.0.0  1.0.0  1  1.2  1.x  ^1.2.3  ~1.2
 * Multiple space/comma separated comparators are ANDed. '||' is unsupported.
 */
export function parseVersionRange(input: string): RangeComparator[] {
	const raw = (input || "").trim();
	if (!raw) throw new VersionRangeError("empty version_range");
	// D1: bound the work an unauthenticated caller can trigger. Without these caps a
	// huge version_range builds an enormous comparator array evaluated against every
	// fetched row. A realistic range is well under these limits.
	if (raw.length > 1000) throw new VersionRangeError("version_range too long (max 1000 chars)");
	if (raw.includes("||")) {
		throw new VersionRangeError("OR ranges ('||') are not supported");
	}
	const tokens = raw.split(/[\s,]+/).filter(Boolean);
	if (tokens.length > 20) throw new VersionRangeError("too many version constraints (max 20)");
	const comparators: RangeComparator[] = [];
	for (const tok of tokens) {
		let m: RegExpExecArray | null;
		if ((m = /^(>=|<=|>|<|=)(.+)$/.exec(tok))) {
			const op = m[1] as RangeComparator["op"];
			const parts = partsOf(m[2].trim());
			if (parts.length === 0) throw new VersionRangeError(`invalid version in "${tok}"`);
			comparators.push({ op, key: keyOf(parts) });
			continue;
		}
		if (tok[0] === "^" || tok[0] === "~") {
			const expanded = expandRange(tok[0], tok.slice(1));
			if (!expanded) throw new VersionRangeError(`invalid range "${tok}"`);
			comparators.push({ op: ">=", key: keyOf(expanded[0]) });
			comparators.push({ op: "<", key: keyOf(expanded[1]) });
			continue;
		}
		// bare token
		const parts = partsOf(tok);
		if (parts.length === 0) throw new VersionRangeError(`invalid version "${tok}"`);
		const hasWildcard = /[xX*]/.test(tok);
		if (parts.length >= 3 && !hasWildcard) {
			// D6: a fully-specified bare token (3+ numeric parts, e.g. nuget "1.0.0.0")
			// is an exact match, consistent with "=1.0.0.0" and normalizeVersion.
			comparators.push({ op: "=", key: keyOf(parts) });
		} else {
			const expanded = expandRange("", tok);
			if (!expanded) throw new VersionRangeError(`invalid range "${tok}"`);
			comparators.push({ op: ">=", key: keyOf(expanded[0]) });
			comparators.push({ op: "<", key: keyOf(expanded[1]) });
		}
	}
	if (comparators.length === 0) throw new VersionRangeError("no comparators parsed");
	return comparators;
}

File: container/version_keys.py
"""
Canonical, self-contained version normalisation for the SBOM dependency index.

╔══════════════════════════════════════════════════════════════════════════╗
║  PARITY CONTRACT                                                          ║
║  This module MUST produce byte-identical sort keys and identical          ║
║  classification to api/src/services/versionKeys.ts. Both are validated    ║
║  against the shared golden vectors in testdata/version-key-vectors.json.  ║
║  Never change one side without the other and without extending vectors.   ║
╚══════════════════════════════════════════════════════════════════════════╝

A *sort key* is an ASCII string whose lexicographic (Unicode code point)
ordering equals the semantic ordering of release versions. Because BigQuery,
Python and JavaScript all compare ASCII strings by code point, the same key
orders identically in all three. Range predicates therefore become plain
string range scans in BigQuery (``WHERE version_sort_key >= @lo AND < @hi``)
with no UDF and no IEEE-754 precision ceiling.

Design notes:
  * Fixed arity / fixed width zero-padding makes release segments compare
    numerically under lexicographic ordering (0.9.2 < 0.10.0).
  * A release sorts ABOVE all of its prereleases: release discriminator is
    '~' (0x7E, highest printable used here); prerelease discriminator starts
    with '-' (0x2D, below digits) so prereleases sort first.
  * Epoch (PEP 440) is a zero-padded prefix so 1!1.0 > 2.0.
  * We deliberately hand-roll a simple parser in BOTH languages rather than
    trust two different third-party libraries to agree byte-for-byte.
"""
from __future__ import annotations

import re
from urllib.parse import unquote

# --- canonical constants (MUST match versionKeys.ts) ---
ARITY = 6          # number of release segments encoded
WIDTH = 10         # zero-pad width per release segment (covers CalVer / epoch-seconds)
SEG_MAX = 10 ** WIDTH - 1
PRE_NUM_WIDTH = 12  # zero-pad width for numeric prerelease identifiers

# Ecosystems whose numeric-dotted release strings compare exactly via the key.
EXACT_SCHEMES = frozenset(
    {"npm", "pypi", "gem", "cargo", "golang", "nuget", "composer", "pub", "swift"}
)
# Ecosystems that are not real installable package releases (self-ref / CI actions).
NON_PACKAGE_ECOSYSTEMS = frozenset({"github", "githubactions"})

_RANGE_PREFIX = ("^", "~", ">", "<")
_GO_PSEUDO_RE = re.compile(r"-\d{12,14}-[0-9a-f]{12}$")
_SHA_RE = re.compile(r"^[0-9a-fA-F]{7,40}$")
_LEADING_NUM_RE = re.compile(r"^(\d+(?:\.\d+)*)")
_PRE_SPLIT_RE = re.compile(r"[.\-_]")


def _result(kind, comparable, exact_scheme, is_prerelease=False, epoch=0, sort_key=None):
    return {
        "version_kind": kind,
        "comparable": comparable,
        "exact_scheme": exact_scheme,
        "is_prerelease": is_prerelease,
        "epoch": epoch,
        "version_sort_key": sort_key,
    }


def _is_range_or_wildcard(v: str) -> bool:
    s = v.strip()
    if s.startswith("="):
        s = s[1:].strip()
    if not s:
        return False
    if s[0] in _RANGE_PREFIX:
        return True
    if "||" in v or "," in v or " - " in v:
        return True
    if "*" in s:
        return True
    for seg in _PRE_SPLIT_RE.split(s):
        if seg in ("x", "X"):
            return True
    return False


def _is_sha(v: str) -> bool:
    if _GO_PSEUDO_RE.search(v):
        return True
    if v.isdigit():
        return False  # pure-digit is CalVer, comparable
    return bool(_SHA_RE.match(v))


def _split_pre(s: str):
    return [t.lower() for t in _PRE_SPLIT_RE.split(s) if t != ""]


def _parse_semver_family(v: str):
    core = v.split("+", 1)[0]
    rel, dash, pre = core.partition("-")
    release = []
    for seg in rel.split("."):
        if not seg.isdigit():
            return None
        release.append(int(seg))
    if not release:
        return None
    is_pre = bool(dash) and pre != ""
    return 0, release, is_pre, (_split_pre(pre) if is_pre else [])


def _parse_pypi(v: str):
    s = v.lower().strip()
    epoch = 0
    if "!" in s:
        e, _, s = s.partition("!")
        epoch = int(e) if e.isdigit() else 0
    s = s.split("+", 1)[0]
    m = _LEADING_NUM_RE.match(s)
    if not m:
        return None
    release = [int(x) for x in m.group(1).split(".")]
    rest = s[m.end():]
    is_pre = False
    pre_tags = []
    if rest and re.match(r"^[._-]?(a|b|c|rc|alpha|beta|pre|preview|dev)", rest):
        is_pre = True
        pre_tags = _split_pre(rest)
    # .postN / unknown trailing -> treated as release (not prerelease)
    return epoch, release, is_pre, pre_tags


def _parse_gem(v: str):
    release = []
    pre_segs = []
    is_pre = False
    for seg in v.split("."):
        if seg.isdigit() and not is_pre:
            release.append(int(seg))
        else:
            is_pre = True
            pre_segs.append(seg)
    if not release:
        return None
    pre_tags = _split_pre(".".join(pre_segs)) if pre_segs else []
    return 0, release, is_pre, pre_tags


def _parse_maven(v: str):
    m = _LEADING_NUM_RE.match(v)
    if not m:
        return None  # codename like Edgware.RELEASE
    release = [int(x) for x in m.group(1).split(".")]
    rest = v[m.end():].lower()
    is_pre = bool(re.search(r"(alpha|beta|rc|m\d|snapshot|cr\d|milestone|preview)", rest))
    return 0, release, is_pre, (_split_pre(rest) if is_pre else [])


def _parse_version(eco: str, v: str):
    if eco in ("npm", "cargo", "golang", "nuget", "composer", "pub", "swift"):
        return _parse_semver_family(v)
    if eco == "pypi":
        return _parse_pypi(v)
    if eco == "gem":
        return _parse_gem(v)
    if eco == "maven":
        return _parse_maven(v)
    # unknown ecosystem: try semver family then pypi-style leading numeric
    parsed = _parse_semver_family(v)
    if parsed is not None:
        return parsed
    return _parse_pypi(v)


def build_sort_key(epoch: int, release, is_prerelease: bool, pre_tags) -> str:
    epoch_part = str(min(max(epoch, 0), 9999)).zfill(4)
    segs = []
    for i in range(ARITY):
        val = release[i] if i < len(release) else 0
        val = min(max(val, 0), SEG_MAX)
        segs.append(str(val).zfill(WIDTH))
    release_part = ".".join(segs)
    if not is_prerelease:
        disc = "~"
    else:
        norm = []
        for t in pre_tags:
            if t.isdigit():
                norm.append("0" + t.zfill(PRE_NUM_WIDTH))
            else:
                norm.append("1" + t)
        disc = "-" + ".".join(norm) if norm else "-"
    return f"{epoch_part}!{release_part}.{disc}"


def normalize_version(ecosystem: str, raw_version) -> dict:
    """Classify and (where comparable) build the sort key for one version string.

    ``ecosystem`` is the PURL type (npm, pypi, ...). ``raw_version`` should be the
    already %-decoded version; we defensively decode again (idempotent).
    """
    eco = (ecosystem or "").lower()
    exact_scheme = eco in EXACT_SCHEMES
    v = unquote(raw_version if raw_version is not None else "").strip()

    if v == "":
        return _result("empty", False, exact_scheme)
    if eco == "github":
        return _result("branch", False, False)
    if eco == "githubactions":
        kind = "wildcard" if _is_range_or_wildcard(v) else "branch"
        return _result(kind, False, False)
    if _is_range_or_wildcard(v):
        return _result("range", False, exact_scheme)

    vv = v
    if len(vv) > 1 and vv[0] in ("v", "V") and vv[1].isdigit():
        vv = vv[1:]
    if vv.startswith("="):
        vv = vv[1:].strip()

    if _is_sha(vv):
        return _result("sha", False, exact_scheme)

    parsed = _parse_version(eco, vv)
    if parsed is None:
        return _result("unparseable", False, exact_scheme)

    epoch, release, is_pre, pre_tags = parsed
    key = build_sort_key(epoch, release, is_pre, pre_tags)
    return _result(
        "prerelease" if is_pre else "release",
        True,
        exact_scheme,
        is_prerelease=is_pre,
        epoch=epoch,
        sort_key=key,
    )


def parse_purl(purl: str):
    """Parse a (possibly %-encoded) PURL into (ecosystem, package_name, version).

    pkg:<type>/<namespace>/<name>@<version>?<qualifiers>#<subpath>
    Returns decoded components. version may be '' when absent.
    """
    if not purl or not purl.startswith("pkg:"):
        return None
    body = purl[4:]
    # strip qualifiers / subpath
    body = body.split("#", 1)[0].split("?", 1)[0]
    # version follows the last '@'; encoded scoped names use %40 so contain no literal '@'
    head, sep, tail = body.rpartition("@")
    if sep:
        name_part, version = head, tail
    else:
        name_part, version = tail, ""
    eco, _, ns_name = name_part.partition("/")
    eco = unquote(eco).lower()
    package_name = unquote(ns_name)
    version = unquote(version)
    return eco, package_name, version


def package_key(ecosystem: str, package_name: str) -> str:
    """Canonical match key for a package within an ecosystem.

    npm: lowercase (scoped names kept). pypi: PEP 503 normalisation
    (lowercase, runs of [._-] -> '-'). Others: lowercase.
    """
    eco = (ecosystem or "").lower()
    name = package_name or ""
    if eco == "pypi":
        return re.sub(r"[-_.]+", "-", name).lower()
    return name.lower()
